feat(scheduler-targets): eventBridge putEvents target

WtfJoke opened the pull request:
An eventBridgePutEvents target was implemented similar to the already existing LambdaInvoke/StepFunctionStartExecution target.

I needed to change some properties and methods from Target.ts from private to protected so that the logic could be reused (hope that is ok).

Some design choices to outline (let me know if you disagree or have improvements I could take :) ):
1. PutEvents would accept multiple Entries (eg. an array), but I decided to support just one single event, because how Target is currently designed (to support only one target arn). It also aligns with the templated integration in the aws management console.
2. It throws an error in the constructor if the base prop `input` is used. All the props should be delivered by the new `EventBridgePutEventsEntry`. It felt not right for the developer experience to split the object (detail to `input` and `source`, `detailType` to `EventBridgePutEventsEntry` ).


Closes #27454.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 3077498c26fbaa273e83201ca961f1ab2e86b7ed
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ed478c73-5e25-4e89-9300-9a1c1499b9f0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a82c9e4e42045535b75a94b66fc0c4fd7fc9244a
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=35fa1e62-5a8b-464b-adf2-93431be30111%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: f55029a7fd7f0f78f88a08a06a10c58a2b95e9f0
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f2b63ad7-e1aa-47ae-a3ae-fd169531ea3d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

kaizencc commented on packages/@aws-cdk/aws-scheduler-targets-alpha/lib/event-bridge-put-events.ts:
```diff
@@ -0,0 +1,107 @@
+ * Send an event to an AWS EventBridge by AWS EventBridge Scheduler.
+ */
+export class EventBridgePutEvents extends ScheduleTargetBase implements IScheduleTarget {
+
```
```suggestion
```

kaizencc commented on packages/@aws-cdk/aws-scheduler-targets-alpha/lib/event-bridge-put-events.ts:
```diff
@@ -0,0 +1,107 @@
+    private readonly props: ScheduleTargetBaseProps,
+  ) {
+    super(props, entry.eventBus.eventBusArn);
+    if (this.baseProps.input) {
+      throw new Error('ScheduleTargetBaseProps.input is not supported for EventBridgePutEvents. Please use entry.detail instead.');
+    }
```
I don't quite understand this even with your pr description. `detail` and `input` seem to both have the same type. Why do we accept one but not the other?

WtfJoke replied:
Theoretically we could accept both (`input` and `detail`). However if somebody is supplying both, which should we pick? 
Then you might think that we can make both optional and throw an exception if both are supplied, but I dont see any advantages going that route.

We could also omit `detail` from `EventBridgePutEventsEntry`, but I think that might be confusing for users, who know the structure from the [PutEvents API Call](https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_PutEvents.html), eg. `detail`. And I think its a bit unintuitive to put most of the payload in one object and the main part (e.g detail in input). So this was my reasoning for the current implementation.

Feel free to question me further or share me how you think about this. :)
I hope I was able to describe it good enough to understand.

kaizencc commented on packages/@aws-cdk/aws-scheduler-targets-alpha/lib/event-bridge-put-events.ts:
```diff
@@ -0,0 +1,107 @@
+    eventBus.grantPutEventsTo(role);
+  }
+
+  protected bindBaseTargetConfig(_schedule: ISchedule): ScheduleTargetConfig {
```
I don't like the duplication here. let's figure out how to modify `super.bindBaseTargetConfig()` so that it can be called via this method and we can reuse as much as we can

WtfJoke replied:
I've introduced a new method overrideTargetConfig which tries to minimize the duplication here. Im open for better ideas, in case it doesnt satisfy your needs :)

kaizencc commented on packages/@aws-cdk/aws-scheduler-targets-alpha/lib/event-bridge-put-events.ts:
```diff
@@ -0,0 +1,107 @@
+    private readonly props: ScheduleTargetBaseProps,
+  ) {
+    super(props, entry.eventBus.eventBusArn);
+    if (this.baseProps.input) {
```
this can also be `this.props.input`, i don't think we need to expose `baseProps` at all

WtfJoke replied:
Yeah, did that

kaizencc commented on packages/@aws-cdk/aws-scheduler-targets-alpha/lib/target.ts:
```diff
@@ -68,7 +68,7 @@ export interface ScheduleTargetBaseProps {
 
   constructor(
-    private readonly baseProps: ScheduleTargetBaseProps,
+    protected readonly baseProps: ScheduleTargetBaseProps,
```
after some of these changes i suspect we can revert this line and the rest of the changes in this file

WtfJoke replied:
Reverted the visibility changes back with the introduction of the method `overrideTargetConfig`

kaizencc reviewed (changes requested):
Hi @WtfJoke! Thanks for another scheduler target! A few comments here and there, but I appreciate your dedication to this module :)

WtfJoke commented on packages/@aws-cdk/aws-scheduler-targets-alpha/README.md:
````diff
@@ -102,3 +102,30 @@ new Schedule(this, 'Schedule', {
+  schedule: ScheduleExpression.rate(Duration.hours(1)),
+  target: new targets.EventBridgePutEvents(eventEntry, {}),
+});
+```
````
StepFunction was already there. I've added eventbridge

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 5acbb445c83f4b00bd00e06960821a7720016cb8
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f8b384de-6d52-479a-b691-244636e9be7a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: e7723b943e4cad6552e4c6c5df863a950cc43a6a
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6fba9a51-6e23-4d6a-a460-01ff28eee9ef%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 7b4f63c729dca7f5a95bd5bf293f493a6cefdc40
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=35ec4351-9a82-4581-b7e7-907815e1efb7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 2ed446044d5f426d4806d70a96588a5a7f36f848
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c15e0f83-89f1-4e1f-b5fd-dcd98415e590%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

WtfJoke commented:
Hey @kaizencc 
Thanks for the review! Appreciate the feedback. I've addressed most of your feedback, please have a look at it again.

Yeah I initially needed StepFunction for a Demo/Talk, so I was keen in contributing this back (after hacking it together on my demo repo) and figured its quiet straight forward to add more targets (Im thinking about adding the Universal target as well, but I could imagine this will be a bit harder) :D

aws-cdk-automation commented:
This PR cannot be merged because it has conflicts. Please resolve them. The PR will be considered stale and closed if it remains in an unmergeable state.

aws-cdk-automation commented:
This PR cannot be merged because it has conflicts. Please resolve them. The PR will be considered stale and closed if it remains in an unmergeable state.

aws-cdk-automation commented:
This PR cannot be merged because it has conflicts. Please resolve them. The PR will be considered stale and closed if it remains in an unmergeable state.

aws-cdk-automation commented:
This PR cannot be merged because it has conflicts. Please resolve them. The PR will be considered stale and closed if it remains in an unmergeable state.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 17938bf0b13740a869bf8dfc9dda3ca8e7ff6b52
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c1604fb8-9378-4865-b5a4-dca4543e34b4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 8a7c82a3854193c27d329b326caec5b4ae9ff209
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8cec4a03-9f4e-4270-a774-1ec0a2a60b64%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

WtfJoke commented:
@kaizencc I've fixed the merge conflicts and rebased my changes. Please have a look at it.

aws-cdk-automation commented:
This PR cannot be merged because it has conflicts. Please resolve them. The PR will be considered stale and closed if it remains in an unmergeable state.

aws-cdk-automation commented:
This PR cannot be merged because it has conflicts. Please resolve them. The PR will be considered stale and closed if it remains in an unmergeable state.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 850638d5eba24e75b81f3f893ccd9a1791108593
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=791c6435-0923-4268-9dc1-ba3d5f63d3d1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

kaizencc commented on packages/@aws-cdk/aws-scheduler-targets-alpha/lib/event-bridge-put-events.ts:
```diff
@@ -0,0 +1,91 @@
+   * The event body
+   *
+   * Can either be provided as an object or as a JSON-serialized string
+   * @example
```
with `@example`, eventally `jsii-rosetta` (a tool we have that aggregates all examples that show up in our documentation) will want this to compile. Unless you want to figure out how to get this to compile, I suggest the following:

```suggestion
   * For example,
   *
   * ```
   * // your uncompiled example code here
   * ```
```

WtfJoke replied:
The build is green as far as I know. Or which build (or command) do you refer to?

kaizencc replied:
It does compile :). Nevermind.

kaizencc commented on packages/@aws-cdk/aws-scheduler-targets-alpha/lib/event-bridge-put-events.ts:
```diff
@@ -0,0 +1,91 @@
+   * For example, events by CloudTrail have detail type "AWS API Call via CloudTrail"
+   * @see https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-events.html
+   */
+  readonly detailType: string;
```
I think we can type this better than a `string`. From the docs you've provided, the only thing that I see as a possible input here is `"AWS API Call via CloudTrail"`. Lol. So here is what I propose:

```ts
readonly detailType: DetailType;

////


export class DetailType {
  public static API_CALL_VIA_CLOUDTRAIL() {
    return new DetailType('AWS API Call via CloudTrail');
  }

  public constructor (public readonly type: string) {}
}
```

This way, we can

a) give users an easy API to hit rather than getting the specific string correct (`DetailType.API_CALL_VIA_CLOUDTRAIL`),
b) provide space for future additions if we know of other valid detail types as static methods,
c) users can still specify whatever they want via `new DetailType('my special type')`

WtfJoke replied:
Will do.

There seems to be [more](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-service-event-list.html), but not documented which `detail-type` they emit.
I found only one additional example in the docs [here](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-events-structure.html) `"detail-type": "EC2 Instance State-change Notification"`

Cant speak for others but our team uses only own detail types (e.g your point c).

For context:
I followed the existing [EventBridgePutEvents](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_stepfunctions_tasks.EventBridgePutEvents.html) implementation of the step function tasks package, which uses string as types.

WtfJoke replied:
Done :)

kaizencc replied:
Andddd I am now wavering on this suggestion. I still think it could be a good idea, but might be overengineered. I didn't see the nuance at first that most `detail-types` are user-owned, and then for events from AWS there maybe detail types with specific strings.

I think we should go back to using `string` here.

WtfJoke replied:
I just dropped my last commit, which effectively reverts the change.

kaizencc reviewed (dismissed):
Hi @WtfJoke, thanks for your continued patience here. I have a few more comments, but this is looking good!

WtfJoke commented on packages/@aws-cdk/aws-scheduler-targets-alpha/lib/event-bridge-put-events.ts:
```diff
@@ -0,0 +1,91 @@
+    private readonly props: ScheduleTargetBaseProps,
+  ) {
+    super(props, entry.eventBus.eventBusArn);
+    if (this.props.input) {
+      throw new Error('ScheduleTargetBaseProps.input is not supported for EventBridgePutEvents. Please use entry.detail instead.');
+    }
```
I got you. However currently its not possible to not set entry.Detail as its required. Do you suggest weakening the type of EventBridgePutEventsEntry.detail so that we can pass null there? And then add this check? It seems to be not ideal.

kaizencc replied:
Ohhh ok got you. This is okay then.

WtfJoke commented:
@kaizencc Great to hear. Looking forward to getting this merged.

Would be glad if you can answer my questions, as im stuck in 2/3 review comments.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 2cd6d3213d2f0c25ab654996c9d83e462c0e3fca
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b18ff2a2-67be-486a-a4ec-b52a20fafde5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

WtfJoke commented:
@kaizencc I just dropped my last commit. Also rebased my changes on to latest main and fixed all the conflicts.

Please have a look :)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ea36560832d428b55410f2fcede6ce5742a9aec0
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b7e021a8-5f5f-4609-9cda-71e1001a0243%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: b28d247d4f2340b5d8fcab02cc4c90d9d133e9aa
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2c674d01-16b4-4cc5-a75c-5fd83e9d1074%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 21e3410bf6a2cf55782e6b0f1359b65754228e16
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=16bb3ed1-71b3-40c6-a5a2-8b1f85a7fc25%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

kaizencc reviewed (approved):
Thanks @WtfJoke and sorry for the many rounds of fruitless reviews :)
